Caddy Webserver for Frontends

ianliuwk1019 opened the pull request:
Caddy webserver for frontends.

Add Caddyfile for serving frontend.
Dockerfile change to build and deploy frontend in container with Caddy.
Migrate content-security-policy settings from previous server.js (using Express) to Caddy.
Verify on browser frontends are running with no issue.
Remove server.js (public/admin) and 'helmet.js' from package.json dependency.

ianliuwk1019 committed:
- Initial local working Caddyfile.

ianliuwk1019 committed:
Merge branch 'main' into feat/caddy-webserver-public-frontend

ianliuwk1019 committed:
Update Caddyfile comments.

ianliuwk1019 committed:
Adjust Caddy root.

ianliuwk1019 committed:
Merge branch 'main' into feat/caddy-webserver-public-frontend

ianliuwk1019 committed:
Add initial docker for running caddy (need more dockerfile adjustment later)

ianliuwk1019 committed:
Merge branch 'main' into feat/caddy-webserver-public-frontend

ianliuwk1019 committed:
Try commenting out USER to see if it resolve the container permission issue: {"level":"error","ts":1658432389.9087143,"msg":"unable to autosave config","file":"/config/caddy/autosave.json","error":"open /config/caddy/autosave.json: permission denied"}

ianliuwk1019 committed:
Try with root user...

ianliuwk1019 committed:
Just some comment clean up and add back USER 1001.

ianliuwk1019 committed:
Try mounting volume for caddy /data

ianliuwk1019 committed:
Try again without USER 1001

ianliuwk1019 committed:
Sigh....sorry I mounted the wrong path. Mount '/config' now.

ianliuwk1019 committed:
Add Caddy default ports.

ianliuwk1019 committed:
Fix deployment frontend links.

ianliuwk1019 committed:
Uncomment USER

ianliuwk1019 committed:
Add admin's Caddyfile to serve up admin frontend.

ianliuwk1019 committed:
Adjust admin Dockerfile to serve from Caddy.
Mount Caddy necessary volumes.

ianliuwk1019 committed:
Add initial migration of cache and content security (CSP) header values from server.js to public frontend.

ianliuwk1019 committed:
content-security-policy does not seem to work on multi-lines.

ianliuwk1019 committed:
Remove unrecognised directive: script-src-attr

ianliuwk1019 committed:
Add initial migration of cache and content security (CSP) header values from server.js to adminfrontend.

ianliuwk1019 committed:
Update content-security-policy.

ianliuwk1019 committed:
Remove unnecessary headers.

ianliuwk1019 committed:
Remove server.js since is replaced and served with Caddy for frontend.
Adjust Dockerfile, no need to install express and helmet individually for server.js anymore.

ianliuwk1019 committed:
Correct Dockerfile

ianliuwk1019 committed:
Adjust Caddy comments.

ianliuwk1019 committed:
Remove dependency 'helmt'.

ianliuwk1019 committed:
Merge branch 'main' into feat/caddy-webserver

# Conflicts:
#	admin/package.json
#	public/package.json

ianliuwk1019 committed:
Remove helmet install from docker-compose.

basilv commented on admin/Caddyfile:
```diff
@@ -0,0 +1,49 @@
+
+:4200 {
+    log {
+        level DEBUG
```
I'm not sure we want to keep debug logging, seems unnecessary.

basilv commented on admin/Caddyfile:
```diff
@@ -0,0 +1,49 @@
+    header {
+        # allow caching for 5 minutes, for refreshing after deployments.
+        Cache-Control "public, max-age=300"
+        Content-Security-Policy "connect-src 'self' https://*.gov.bc.ca http://*.gov.bc.ca;form-action 'self';img-src 'self' https://server.arcgisonline.com data: https://services.arcgisonline.com data: https://maps.gov.bc.ca;frame-src 'self' https://*.gov.bc.ca;default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;frame-ancestors 'self';object-src 'none';script- ...
```
You could add a comment explaining that the header definition has to remain on a single line.

basilv commented on admin/Caddyfile:
```diff
@@ -0,0 +1,49 @@
+    file_server
+}
+
+##################
```
Probably don't need to keep all these comments, or merge into content above. Derek can make the final call if you are unsure.

basilv commented on admin/Dockerfile:
```diff
@@ -10,11 +10,11 @@ RUN cd libs && \
 # Deployment container
-FROM registry.access.redhat.com/ubi8/ubi-micro
+# FROM registry.access.redhat.com/ubi8/ubi-micro
+FROM caddy:2
```
Derek will want to try to continue to use ubi-micro as the base, but then this means installing Caddy manually, you might want to leave this up to Derek to sort out.

DerekRoberts replied:
@basilv Thanks, agreed!  Ian and I have decided to change the base later.  For now, this is bang on.  :)

basilv commented on admin/Dockerfile:
```diff
@@ -32,14 +32,13 @@ COPY --from=builder /usr/lib64/libbrotlicommon.so.1 /usr/lib64/
 
 # Expose port - mostly a convention, for readability
-EXPOSE 4200
+EXPOSE 4200 2019 80 443
```
Not sure why other ports are exposed. I favor exposing the minimum required.

basilv commented on public/Caddyfile:
```diff
@@ -0,0 +1,48 @@
+# Caddy's configuration file to serve up frontend. (Require Caddy installed/downloaded in system)
+# Local: Assume local build is ready at './dist/public'; at root working directory run command "caddy run --watch" at console.
```
Similar comments as for admin Caddyfile.

basilv commented on public/Dockerfile:
```diff
@@ -10,11 +10,11 @@ RUN cd libs && \
     cd ../public && \
     npm ci && \
```
Similar comments as for admin/Dockerfile

ianliuwk1019 committed:
Try removing port 2019(admin port), 80/443(default ports) and see if Caddy works.
Adjust Caddy log level to ERROR.
Move a comment for content-security-policy close to the directive.

ianliuwk1019 committed:
Remove some comments from Caddyfile.

ianliuwk1019 committed:
Add 1 Mi PVC for Caddy /data

ianliuwk1019 committed:
Increase Caddy /data PVC to 20Mi.

ianliuwk1019 committed:
Try again with higher size.

DerekRoberts reviewed (approved):
Good job!  Lots done.  :)
